handler/src/main/java/io/netty/handler/ssl/DefaultOpenSslSession.java

```diff
@@ -0,0 +1,376 @@
+                values = this.values = new HashMap<String, Object>(2);
+            }
+            old = values.put(name, value);
+        }
+
+        if (value instanceof SSLSessionBindingListener) {
+            ((SSLSessionBindingListener) value).valueBound(new SSLSessionBindingEvent(this, name));
+        }
+        notifyUnbound(old, name);
+    }
+
+    @Override
+    public Object getValue(String name) {
+        ObjectUtil.checkNotNull(name, "name");
+        synchronized (this) {
+            if (values == null) {
+                return null;
+            }
+            return values.get(name);
+        }
+    }
+
+    @Override
+    public void removeValue(String name) {
+        ObjectUtil.checkNotNull(name, "name");
+
+        final Object old;
+        synchronized (this) {
+            Map<String, Object> values = this.values;
+            if (values == null) {
+                return;
+            }
+            old = values.remove(name);
+        }
+
+        notifyUnbound(old, name);
+    }
+
+    @Override
+    public String[] getValueNames() {
+        synchronized (this) {
+            Map<String, Object> values = this.values;
+            if (values == null || values.isEmpty()) {
+                return EmptyArrays.EMPTY_STRINGS;
+            }
+            return values.keySet().toArray(EmptyArrays.EMPTY_STRINGS);
+        }
+    }
+
+    private void notifyUnbound(Object value, String name) {
+        if (value instanceof SSLSessionBindingListener) {
+            ((SSLSessionBindingListener) value).valueUnbound(new SSLSessionBindingEvent(this, name));
+        }
+    }
+
+    @Override
+    public String getPeerHost() {
+        return peerHost;
+    }
+
+    @Override
+    public int getPeerPort() {
+        return peerPort;
+    }
+
+    @Override
+    public OpenSslSessionId sessionId() {
+        return id;
+    }
+
+    @Override
+    public byte[] getId() {
+        return sessionId().cloneBytes();
+    }
+
+    @Override
+    public int hashCode() {
+        return sessionId().hashCode();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (!(obj instanceof OpenSslSession)) {
+            return false;
+        }
+        return sessionId().equals(((OpenSslSession) obj).sessionId());
+    }
+
+    @Override
+    public boolean isNullSession() {
+        return false;
+    }
+
+    @Override
+    public long getCreationTime() {
+        return creationTime;
+    }
+
+    @Override
+    public long getLastAccessedTime() {
+        return lastAccessed;
+    }
+
+    @Override
+    public void invalidate() {
+        if (INVALID_UPDATER.compareAndSet(this, 0, 1)) {
+            sessionContext.removeFromCache(this);
+        }
+    }
+
+    @Override
+    public boolean isValid() {
+       return isValid(System.currentTimeMillis());
+    }
+
+    @Override
+    public boolean isValid(long now) {
+        if (sslSession == -1 || INVALID_UPDATER.get(this) == 1) {
```

Comment:
We can just read the volatile field directly here right? I was under the impression that one is required to use the updater for writing and CAS operations.

Reply:
yeah ... that said I like to be consistent and just use it always